cpp/include/raft/core/operators.hpp

```diff
@@ -0,0 +1,354 @@
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+struct argmax_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+template <typename ScalarT>
+struct const_op {
+  const ScalarT scalar;
+
+  constexpr explicit const_op(const ScalarT& s) : scalar{s} {}
+
+  template <typename... Args>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(Args...) const
+  {
+    return scalar;
+  }
+};
+
+/**
+ * @brief Wraps around a binary operator, passing a constant on the right-hand side.
+ *
+ * Usage example:
+ * @code{.cpp}
+ *  #include <raft/core/operators.hpp>
+ *
+ *  raft::plug_const_op<float, raft::mul_op> op(2.0f);
+ *  std::cout << op(2.1f) << std::endl;  // 4.2
+ * @endcode
+ *
+ * @tparam ConstT
+ * @tparam BinaryOpT
+ */
+template <typename ConstT, typename BinaryOpT>
+struct plug_const_op {
+  const ConstT c;
+  const BinaryOpT composed_op;
+
+  template <typename OpT     = BinaryOpT,
+            typename UnusedT = std::enable_if_t<std::is_default_constructible_v<OpT>>>
+  constexpr explicit plug_const_op(const ConstT& s)
+    : c{s}, composed_op{}  // The compiler complains if composed_op is not initialized explicitly
+  {
+  }
+  constexpr plug_const_op(const ConstT& s, BinaryOpT o) : c{s}, composed_op{o} {}
+
+  template <typename InT>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(InT a) const
+  {
+    return composed_op(a, c);
+  }
+};
+
+template <typename Type>
+using add_const_op = plug_const_op<Type, add_op>;
+
+template <typename Type>
+using sub_const_op = plug_const_op<Type, sub_op>;
+
+template <typename Type>
+using mul_const_op = plug_const_op<Type, mul_op>;
+
+template <typename Type>
+using div_const_op = plug_const_op<Type, div_op>;
+
+template <typename Type>
+using div_checkzero_const_op = plug_const_op<Type, div_checkzero_op>;
+
+template <typename Type>
+using pow_const_op = plug_const_op<Type, pow_op>;
+
+/**
+ * @brief Constructs an operator by composing a chain of operators.
+ *
+ * Note that all arguments are passed to the innermost operator.
+ *
+ * Usage example:
+ * @code{.cpp}
+ *  #include <raft/core/operators.hpp>
+ *
+ *  auto op = raft::compose_op(raft::sqrt_op(), raft::abs_op(), raft::cast_op<float>(),
+ *                             raft::add_const_op<int>(8));
+ *  std::cout << op(-50) << std::endl;  // 6.48074
+ * @endcode
+ *
+ * @tparam OpsT Any number of operation types.
+ */
+template <typename... OpsT>
+struct compose_op {
+  const std::tuple<OpsT...> ops;
+
+  template <typename TupleT = std::tuple<OpsT...>,
+            typename CondT  = std::enable_if_t<std::is_default_constructible_v<TupleT>>>
+  constexpr compose_op()
```

Comment:
It's just occurred to me that the two constructors would clash if we try to initialize `compose_op` with zero argument operators. For the sake of completeness it could be a proper identity operator otherwise. Do we really need the default constructor here at all?

Reply:
I don't think that zero operators is a valid use case for this structure, and I think that this default constructor adds value. We could maybe add a `static_assert` for the sake of clear error reporting...

Reply:
We're surely talking about very hypothetical things now, but one could imagine calling this struct by another template function in the user code, where zero arguments (functions) means identity. It makes a good monoid of unary functions :)

Reply:
Surprisingly, the structure seems to compile just fine when `Args...` is empty. The second constructor is used. But if I comment out the second constructor, the default one is used. It doesn't see this as multiple definitions.

Reply:
According to the [Overload resolution](https://en.cppreference.com/w/cpp/language/overload_resolution) rules, section _Best viable function_:

> 4) or, if not that, F1 is a non-template function while F2 is a template specialization

The non-template constructor unambiguously takes precedence over the template one.